pull_request_migration/gerald_design.md

```diff
@@ -0,0 +1,79 @@
+# Design proposal for Gerald
+
+Gerald is a user-customizable notification tool for GitHub, similar to what
+[Herald](https://secure.phabricator.com/book/phabricator/article/herald/) does
+for [Phabricator](https://www.phacility.com/phabricator/).
+
+## Vision
+
+Elaine is working on the MLIR sub-project in the LLVM monorepo. She is
+interested in all changes that affect her work. TO get notified on relevant
+changes, she configures a couple of rules in the `.gerald` file in the monorepo:
+
+```yaml
+Auto-Label:
+# Rules that automatically label Pull Requests based on the files they modify
+- path: /llvm/** # path expression 
+  label: llvm # label to assign
+  branches: # notify on commits to these branches 
+  - main
+  - release*
+- path: /mlir/**
+  label: mlir
+  branches:
+  - main
+  - release*
+
+Notifications:
+# Notify users on new labels: When one of these labels is assigned to an issue
+# or Pull Request, notify the list of users
+- labels: 
+  - mlir
+  - llvm  
+  users: 
+  - Elaine1234 # Elaine's github user account
+  - Gao1234 # Elaine's colleague who also wants the same notifications
+
+Email-Mapping:
+# For sending email notifications, we actually need to know the users email 
+# addresses. Github does not give out this.
+- Elaine1234: elainesemail@provider.xyz
+```
+
+Whenever a new PullRequest (PR) is opened Gerald first checks the PR
+against this config file. If an `Auto-Label` rule matches, Gerald assigns this
+label to the PR. In the second phase Gerald checks the `Notifications` rules
+against new labels that were assigned to Issues or PRs. For all matches, Gerald
+adds a comment:
+
+> *GeraldBot commented on Aug 27th:*
+>
+> notifying @Elaine1234, @Gao1234
+
```

Comment:
This is pretty good already, but I fear that the first email will look like:
> @gerald[bot] requested your review on: #1234 <PR title>

With nothing more, right?